plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs

```diff
@@ -1800,20 +1812,31 @@ instance uni ~ DefaultUni => ToBuiltinMeaning uni DefaultFun where
             (runCostingFunOneArgument . paramBlake2b_224)
 
     -- Conversions
+    {- See Note [Input length limitation for IntegerToByteString] -}
     toBuiltinMeaning _semvar IntegerToByteString =
-      let integerToByteStringDenotation :: Bool -> Integer -> Integer -> BuiltinResult BS.ByteString
-          integerToByteStringDenotation = integerToByteStringWrapper
+      let integerToByteStringDenotation :: Bool -> LiteralByteSize -> Integer -> BuiltinResult BS.ByteString
+          {- The second argument is wrapped in a LiteralByteSize to allow us to interpret it as a size during
+             costing.  It appears as an integer in UPLC: see Note [Integral types as Integer]. -}
+          integerToByteStringDenotation b (LiteralByteSize w) n =
+              if w > integerToByteStringMaximumOutputLength ||
+                     (w == 0 &&
+                            8 * fromSatInt (sumCostStream . flattenCostRose . memoryUsage $ n)
```

Comment:
That is a really weird line... The denotation of a builtin depends on costing for one of the arguments? Can we not rely on costing here and instead having some direct byte-handling logic? What if we change the size metric or something.

Reply:
For [other reviewers](https://github.com/IntersectMBO/plutus/pull/5733#issuecomment-1903468249), see this comment later.